gobblin-core/src/main/java/gobblin/source/extractor/extract/google/GoogleCommon.java

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright (C) 2014-2016 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+package gobblin.source.extractor.extract.google;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.net.Proxy;
+import java.net.URI;
+import java.nio.file.Files;
+import java.nio.file.attribute.PosixFilePermissions;
+import java.security.GeneralSecurityException;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.permission.FsAction;
+import org.apache.hadoop.fs.permission.FsPermission;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.api.client.auth.oauth2.Credential;
+import com.google.api.client.googleapis.GoogleUtils;
+import com.google.api.client.googleapis.auth.oauth2.GoogleCredential;
+import com.google.api.client.googleapis.javanet.GoogleNetHttpTransport;
+import com.google.api.client.http.HttpTransport;
+import com.google.api.client.http.javanet.NetHttpTransport;
+import com.google.api.client.json.JsonFactory;
+import com.google.api.client.json.gson.GsonFactory;
+import com.google.common.base.Preconditions;
+
+import static gobblin.configuration.ConfigurationKeys.*;
+import static gobblin.source.extractor.extract.google.GoogleCommonKeys.*;
+import gobblin.configuration.State;
+
+/**
+ * Utility class that has static methods for Google services.
+ *
+ */
+public class GoogleCommon {
+  private static final Logger LOG = LoggerFactory.getLogger(GoogleCommon.class);
+  private static final JsonFactory JSON_FACTORY = GsonFactory.getDefaultInstance();
+  private static final String JSON_FILE_EXTENSION = ".json";
+
+  private static final FsPermission USER_READ_PERMISSION_ONLY = new FsPermission(FsAction.READ,
+                                                                                 FsAction.NONE,
+                                                                                 FsAction.NONE);
+
+  /**
+   * Provides Google credential either via JSON file or via P12 file.
+   * Note that it distinguishes file by extension (e.g: ".json")
+   *
+   * @param state
+   * @param privateKeyFile
+   * @return
+   */
+  public static Credential newSourceCredential(State state) {
```

Comment:
I see. Actually, it was my first approach to use explicit parameters but switched it state because it became too many. Let me refactor to use explicit parameters and make it neat as well.